Duplicate primary key generated for PeristentLog, fails processing

The Id generation strategy for `PersistentLogDAO` is defaulting to `AUTO`, which is generating duplicate ID's in some PersistentLog implementations. 

The PK constraint failure is causing a TX rollbackfor some data processing operations, and therefore some operations are failing to process data, e.g. recalculation of datapoints for change detection

Server side error log:

```
2023-10-31 12:37:01,939 ERROR [io.hyp.too.hor.bus.TaskQueue] (executor-thread-7935) Error executing task in the queue for test 254: org.hibernate.exception.ConstraintViolationException: could not execute statement [ERROR: duplicate key v
alue violates unique constraint "calculationlog_pkey"] [insert into DatasetLog (dataset_id,level,message,source,testid,timestamp,id) values (?,?,?,?,?,?,?)]
        at org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:95)
        at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:56)
        at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:108)
        at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:278)
        at org.hibernate.engine.jdbc.mutation.internal.AbstractMutationExecutor.performNonBatchedMutation(AbstractMutationExecutor.java:107)
        at org.hibernate.engine.jdbc.mutation.internal.MutationExecutorSingleNonBatched.performNonBatchedOperations(MutationExecutorSingleNonBatched.java:40)
        at org.hibernate.engine.jdbc.mutation.internal.AbstractMutationExecutor.execute(AbstractMutationExecutor.java:52)
        at org.hibernate.persister.entity.mutation.InsertCoordinator.doStaticInserts(InsertCoordinator.java:169)
        at org.hibernate.persister.entity.mutation.InsertCoordinator.coordinateInsert(InsertCoordinator.java:111)
        at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:2784)
        at org.hibernate.action.internal.EntityInsertAction.execute(EntityInsertAction.java:102)
        at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:635)
        at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:502)
        at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:358)
        at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:39)
        at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:127)
        at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1412)
        at org.hibernate.internal.SessionImpl.flush(SessionImpl.java:1398)
        at org.hibernate.query.sql.internal.NativeQueryImpl.prepareForExecution(NativeQueryImpl.java:599)
        at org.hibernate.query.spi.AbstractSelectionQuery.beforeQuery(AbstractSelectionQuery.java:391)
        at org.hibernate.query.spi.AbstractSelectionQuery.list(AbstractSelectionQuery.java:364)
        at org.hibernate.query.Query.getResultList(Query.java:119)
        at io.hyperfoil.tools.horreum.svc.AlertingServiceImpl.emitDatapoints(AlertingServiceImpl.java:422)
        at io.hyperfoil.tools.horreum.svc.AlertingServiceImpl.recalculateDatapointsForDataset(AlertingServiceImpl.java:291)
       at io.hyperfoil.tools.horreum.svc.AlertingServiceImpl.recalculateForDataset(AlertingServiceImpl.java:981)
        at io.hyperfoil.tools.horreum.svc.AlertingServiceImpl_Subclass.recalculateForDataset$$superforward(Unknown Source)
        at io.hyperfoil.tools.horreum.svc.AlertingServiceImpl_Subclass$$function$$22.apply(Unknown Source)
        at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:73)
        at io.quarkus.arc.impl.AroundInvokeInvocationContext$NextAroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:97)
        at io.hyperfoil.tools.horreum.server.RolesInterceptor.intercept(RolesInterceptor.java:70)
        at io.hyperfoil.tools.horreum.server.RolesInterceptor_Bean.intercept(Unknown Source)
        at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
        at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:70)
        at io.quarkus.arc.impl.AroundInvokeInvocationContext$NextAroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:97)
        at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:136)
        at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:107)
        at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequiresNew.doIntercept(TransactionalInterceptorRequiresNew.java:41)
        at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.intercept(TransactionalInterceptorBase.java:61)
        at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequiresNew.intercept(TransactionalInterceptorRequiresNew.java:32)
        at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequiresNew_Bean.intercept(Unknown Source)
        at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
        at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:70)
        at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:62)
        at io.hyperfoil.tools.horreum.server.BaseTransactionRetryInterceptor.intercept(BaseTransactionRetryInterceptor.java:29)
        at io.hyperfoil.tools.horreum.server.BaseTransactionRetryInterceptor_RequiresNewTransactionRetryInterceptor_Bean.intercept(Unknown Source)
        at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
        at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:30)
        at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:27)
        at io.hyperfoil.tools.horreum.svc.AlertingServiceImpl_Subclass.recalculateForDataset(Unknown Source)
        at io.hyperfoil.tools.horreum.svc.AlertingServiceImpl.startRecalculation(AlertingServiceImpl.java:943)
        at io.hyperfoil.tools.horreum.svc.AlertingServiceImpl.lambda$recalculateDatapoints$44(AlertingServiceImpl.java:916)
        at io.smallrye.context.impl.wrappers.SlowContextualRunnable.run(SlowContextualRunnable.java:19)
        at io.hyperfoil.tools.horreum.svc.Util.lambda$wrapForBlockingExecution$3(Util.java:496)
        at io.hyperfoil.tools.horreum.bus.TaskQueue.executeOrAdd(TaskQueue.java:27)
        at io.hyperfoil.tools.horreum.bus.MessageBus.lambda$executeForTest$9(MessageBus.java:296)
        at io.vertx.core.impl.ContextBase.lambda$executeBlocking$1(ContextBase.java:180)
        at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:277)
        at io.vertx.core.impl.ContextBase.lambda$internalExecuteBlocking$2(ContextBase.java:199)
        at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "calculationlog_pkey"
        at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2713)
        at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2401)
        at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:368)
        at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:498)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:415)
        at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:190)
        at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:152)
        at io.agroal.pool.wrapper.PreparedStatementWrapper.executeUpdate(PreparedStatementWrapper.java:88)
        at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:275)
        ... 62 more

```

--- a/horreum-backend/src/main/java/io/hyperfoil/tools/horreum/entity/ActionLogDAO.java
+++ b/horreum-backend/src/main/java/io/hyperfoil/tools/horreum/entity/ActionLogDAO.java
@@ -1,10 +1,27 @@
 package io.hyperfoil.tools.horreum.entity;
 
 import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
 import jakarta.validation.constraints.NotNull;
+import org.hibernate.annotations.GenericGenerator;
+import org.hibernate.id.enhanced.SequenceStyleGenerator;
 
 @Entity(name = "ActionLog")
 public class ActionLogDAO extends PersistentLogDAO {
+
+   @Id
+   @GenericGenerator(
+           name = "actionlog_id_generator",
+           strategy = "io.hyperfoil.tools.horreum.entity.SeqIdGenerator",
+           parameters = {
+                   @org.hibernate.annotations.Parameter(name = SequenceStyleGenerator.INCREMENT_PARAM, value = "1"),
+           }
+   )
+   @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "actionlog_id_generator")
+   public Long id;
+
    @NotNull
    public int testId;
 

--- a/horreum-backend/src/main/java/io/hyperfoil/tools/horreum/entity/PersistentLogDAO.java
+++ b/horreum-backend/src/main/java/io/hyperfoil/tools/horreum/entity/PersistentLogDAO.java
@@ -19,10 +19,6 @@ public abstract class PersistentLogDAO extends PanacheEntityBase {
    public static final int WARN = 2;
    public static final int ERROR = 3;
 
-   @Id
-   @GeneratedValue
-   public Long id;
-
    @NotNull
    public int level;
 

--- a/horreum-backend/src/main/java/io/hyperfoil/tools/horreum/entity/alerting/DatasetLogDAO.java
+++ b/horreum-backend/src/main/java/io/hyperfoil/tools/horreum/entity/alerting/DatasetLogDAO.java
@@ -1,17 +1,14 @@
 package io.hyperfoil.tools.horreum.entity.alerting;
 
-import jakarta.persistence.ConstraintMode;
-import jakarta.persistence.Entity;
-import jakarta.persistence.FetchType;
-import jakarta.persistence.ForeignKey;
-import jakarta.persistence.JoinColumn;
-import jakarta.persistence.ManyToOne;
+import jakarta.persistence.*;
 import jakarta.validation.constraints.NotNull;
 
 import io.hyperfoil.tools.horreum.entity.PersistentLogDAO;
 import io.hyperfoil.tools.horreum.entity.data.DatasetDAO;
 import io.hyperfoil.tools.horreum.entity.data.RunDAO;
 import io.hyperfoil.tools.horreum.entity.data.TestDAO;
+import org.hibernate.annotations.GenericGenerator;
+import org.hibernate.id.enhanced.SequenceStyleGenerator;
 
 /**
  * This table is meant to host logged events with relation to {@link DatasetDAO datasets},
@@ -19,6 +16,17 @@ import io.hyperfoil.tools.horreum.entity.data.TestDAO;
  */
 @Entity(name = "DatasetLog")
 public class DatasetLogDAO extends PersistentLogDAO {
+
+   @Id
+   @GenericGenerator(
+           name = "datasetlog_id_generator",
+           strategy = "io.hyperfoil.tools.horreum.entity.SeqIdGenerator",
+           parameters = {
+                   @org.hibernate.annotations.Parameter(name = SequenceStyleGenerator.INCREMENT_PARAM, value = "1"),
+           }
+   )
+   @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "datasetlog_id_generator")
+   public Long id;
 
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "testid", foreignKey = @ForeignKey(ConstraintMode.NO_CONSTRAINT))

--- a/horreum-backend/src/main/java/io/hyperfoil/tools/horreum/entity/alerting/TransformationLogDAO.java
+++ b/horreum-backend/src/main/java/io/hyperfoil/tools/horreum/entity/alerting/TransformationLogDAO.java
@@ -1,19 +1,26 @@
 package io.hyperfoil.tools.horreum.entity.alerting;
 
 
-import jakarta.persistence.ConstraintMode;
-import jakarta.persistence.Entity;
-import jakarta.persistence.FetchType;
-import jakarta.persistence.ForeignKey;
-import jakarta.persistence.JoinColumn;
-import jakarta.persistence.ManyToOne;
+import jakarta.persistence.*;
 
 import io.hyperfoil.tools.horreum.entity.PersistentLogDAO;
 import io.hyperfoil.tools.horreum.entity.data.RunDAO;
 import io.hyperfoil.tools.horreum.entity.data.TestDAO;
+import org.hibernate.annotations.GenericGenerator;
+import org.hibernate.id.enhanced.SequenceStyleGenerator;
 
 @Entity(name = "TransformationLog")
 public class TransformationLogDAO extends PersistentLogDAO {
+   @Id
+   @GenericGenerator(
+           name = "transformationlog_id_generator",
+           strategy = "io.hyperfoil.tools.horreum.entity.SeqIdGenerator",
+           parameters = {
+                   @org.hibernate.annotations.Parameter(name = SequenceStyleGenerator.INCREMENT_PARAM, value = "1"),
+           }
+   )
+   @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "transformationlog_id_generator")
+   public Long id;
 
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "testid", foreignKey = @ForeignKey(ConstraintMode.NO_CONSTRAINT))

--- a/horreum-backend/src/main/java/io/hyperfoil/tools/horreum/entity/report/ReportLogDAO.java
+++ b/horreum-backend/src/main/java/io/hyperfoil/tools/horreum/entity/report/ReportLogDAO.java
@@ -1,13 +1,23 @@
 package io.hyperfoil.tools.horreum.entity.report;
 
-import jakarta.persistence.Entity;
-import jakarta.persistence.JoinColumn;
-import jakarta.persistence.ManyToOne;
+import jakarta.persistence.*;
 
 import io.hyperfoil.tools.horreum.entity.PersistentLogDAO;
+import org.hibernate.annotations.GenericGenerator;
+import org.hibernate.id.enhanced.SequenceStyleGenerator;
 
 @Entity(name = "ReportLog")
 public class ReportLogDAO extends PersistentLogDAO {
+   @Id
+   @GenericGenerator(
+           name = "reportlog_id_generator",
+           strategy = "io.hyperfoil.tools.horreum.entity.SeqIdGenerator",
+           parameters = {
+                   @org.hibernate.annotations.Parameter(name = SequenceStyleGenerator.INCREMENT_PARAM, value = "1"),
+           }
+   )
+   @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "reportlog_id_generator")
+   public Long id;
    @ManyToOne(optional = false)
    @JoinColumn(name = "report_id")
    TableReportDAO report;

--- a/horreum-backend/src/main/resources/db/changeLog.xml
+++ b/horreum-backend/src/main/resources/db/changeLog.xml
@@ -4180,4 +4180,18 @@
             DROP TABLE IF EXISTS messagebus;
         </sql>
     </changeSet>
+
+    <changeSet id="111" author="johara">
+        <createSequence sequenceName="actionlog_id_generator" startValue="1" incrementBy="1" cacheSize="1" />
+        <createSequence sequenceName="datasetlog_id_generator" startValue="1" incrementBy="1" cacheSize="1" />
+        <createSequence sequenceName="reportlog_id_generator" startValue="1" incrementBy="1" cacheSize="1" />
+        <createSequence sequenceName="transformationlog_id_generator" startValue="1" incrementBy="1" cacheSize="1" />
+        <sql>
+            select setval('actionlog_id_generator', (select max(id)+1 from actionlog), false);
+            select setval('datasetlog_id_generator', (select max(id)+1 from datasetlog), false);
+            select setval('reportlog_id_generator', (select max(id)+1 from reportlog), false);
+            select setval('transformationlog_id_generator', (select max(id)+1 from transformationlog), false);
+            GRANT ALL ON SEQUENCE actionlog_id_generator, datasetlog_id_generator, reportlog_id_generator, transformationlog_id_generator TO "${quarkus.datasource.username}";
+        </sql>
+    </changeSet>
 </databaseChangeLog>

